File: src/main/java/dev/ergenverse/runtime/layer/BlueprintLayer.java
package dev.ergenverse.runtime.layer;

import dev.ergenverse.runtime.PlanetSuzakuBlueprint;
import dev.ergenverse.runtime.Provenance;
import dev.ergenverse.runtime.spatial.SpatialIndex;

/**
 * BlueprintLayer — the immutable canon stratum (Provenance.CANON).
 *
 * <p><b>Architectural directive (CRON-69, point 8):</b> "PlanetSuzakuBlueprint
 * should never answer 'getBlock'. … Instead I'd make the blueprint answer
 * higher-level questions. Like queryTerrain(...), queryStructures(...),
 * queryActors(...), querySpiritVeins(...). Why? Because the blueprint isn't
 * actually a giant block array. It's a description of the world."
 *
 * <p>So this layer's {@link #getBlock} <b>always returns null</b> — it defers
 * to the minecraft:noise base terrain (which is itself deterministic via
 * {@code DeterministicSeedHandler.CANON_SEED}, so "zero dependence on random
 * terrain" still holds). The blueprint never pretends to be a per-block array.
 *
 * <p>Instead, {@link #getChunkContribution} answers at <b>structure granularity</b>
 * (point 7: "Blueprint returns Structures, Terrain patches, Road segments"):
 * it queries the {@link SpatialIndex} for every {@link PlanetSuzakuBlueprint.CanonLocation}
 * whose footprint intersects the chunk. The {@link dev.ergenverse.runtime.materialize.ChunkMaterializer}
 * then invokes each location's registered builder (idempotent) to place the
 * hand-authored blocks. That is how canon enters the live world — as named
 * structures materialized on demand, never as a per-block lookup.
 *
 * <p>MC 1.20.1 / Forge 47.4.0 / Java 17.</p>
 */
public final class BlueprintLayer implements WorldLayer {

    private final PlanetSuzakuBlueprint blueprint;
    private final SpatialIndex<PlanetSuzakuBlueprint.CanonLocation> spatialIndex;

    public BlueprintLayer(PlanetSuzakuBlueprint blueprint,
                          SpatialIndex<PlanetSuzakuBlueprint.CanonLocation> spatialIndex) {
        this.blueprint = blueprint;
        this.spatialIndex = spatialIndex;
    }

    @Override public Provenance provenance() { return Provenance.CANON; }

    /**
     * Always null. The blueprint is a description of the world, not a block
     * array. Per-block canon is materialized by builders via
     * {@link #getChunkContribution}, not by this method.
     */
    @Override
    public String getBlock(int x, int y, int z) {
        return null;
    }

    /**
     * The canon structures (settlements, sects, ruins, geographic landmarks)
     * whose footprint intersects this chunk. The materializer builds each via
     * its registered {@code StructureBuilder} (idempotent).
     */
    @Override
    public ChunkContribution getChunkContribution(int chunkX, int chunkZ) {
        ChunkContribution c = new ChunkContribution();
        int minX = chunkX * 16, minZ = chunkZ * 16;
        int maxX = minX + 15, maxZ = minZ + 15;
        for (PlanetSuzakuBlueprint.CanonLocation loc : spatialIndex.query(minX, minZ, maxX, maxZ)) {
            c.structures.add(loc);
        }
        return c;
    }

    /** The blueprint this layer reads from. */
    public PlanetSuzakuBlueprint blueprint() { return blueprint; }
}

File: src/engine/world/BlueprintLayer.ts
/**
 * BlueprintLayer — port of dev.ergenverse.runtime.layer.BlueprintLayer
 *
 * The blueprint is a DESCRIPTION, not a per-block truth. getBlock always
 * returns null — it answers higher-level queries (queryStructures,
 * queryActors, querySpiritVeins, queryLandmarks).
 *
 * Point 8 of the CRON-69 ten-point architecture: "The blueprint never
 * answers getBlock — it answers higher-level queries."
 *
 * The blueprint is NEVER modified. Every new save starts from identical
 * canon; divergence lives in the WorldDeltaStore journal.
 */
import type { WorldLayer, BlockQueryResult, StructureBox } from './WorldLayer'
import { Provenance } from './Provenance'
import type { CanonActor, SpiritVein, Landmark } from '../canon/types'

export class BlueprintLayer implements WorldLayer {
  readonly provenance = Provenance.CANON

  private readonly structures: StructureBox[] = []
  private readonly actors: CanonActor[] = []
  private readonly spiritVeins: SpiritVein[] = []
  private readonly landmarks: Landmark[] = []

  registerStructure(s: StructureBox): void {
    this.structures.push(s)
  }
  registerActor(a: CanonActor): void {
    this.actors.push(a)
  }
  registerSpiritVein(v: SpiritVein): void {
    this.spiritVeins.push(v)
  }
  registerLandmark(l: Landmark): void {
    this.landmarks.push(l)
  }

  /** ALWAYS null. Blueprint never answers per-block. */
  getBlock(_x: number, _y: number, _z: number): BlockQueryResult {
    return null
  }

  queryStructures(chunkX: number, chunkZ: number): StructureBox[] {
    const minCx = chunkX * 16
    const minCz = chunkZ * 16
    const maxCx = minCx + 15
    const maxCz = minCz + 15
    return this.structures.filter(
      (s) =>
        s.maxX >= minCx &&
        s.minX <= maxCx &&
        s.maxZ >= minCz &&
        s.minZ <= maxCz,
    )
  }

  queryActors(chunkX: number, chunkZ: number): CanonActor[] {
    const minCx = chunkX * 16
    const minCz = chunkZ * 16
    const maxCx = minCx + 15
    const maxCz = minCz + 15
    return this.actors.filter((a) => {
      const x = a.position[0]
      const z = a.position[2]
      return x >= minCx && x <= maxCx && z >= minCz && z <= maxCz
    })
  }

  querySpiritVeins(chunkX: number, chunkZ: number): SpiritVein[] {
    const minCx = chunkX * 16
    const minCz = chunkZ * 16
    const maxCx = minCx + 15
    const maxCz = minCz + 15
    return this.spiritVeins.filter((v) => {
      const x = v.position[0]
      const z = v.position[2]
      return x >= minCx && x <= maxCx && z >= minCz && z <= maxCz
    })
  }

  queryLandmarks(chunkX: number, chunkZ: number): Landmark[] {
    const minCx = chunkX * 16
    const minCz = chunkZ * 16
    const maxCx = minCx + 15
    const maxCz = minCz + 15
    return this.landmarks.filter((l) => {
      const x = l.position[0]
      const z = l.position[2]
      return x >= minCx && x <= maxCx && z >= minCz && z <= maxCz
    })
  }

  allStructures(): StructureBox[] {
    return [...this.structures]
  }
  allActors(): CanonActor[] {
    return [...this.actors]
  }
  allSpiritVeins(): SpiritVein[] {
    return [...this.spiritVeins]
  }
  allLandmarks(): Landmark[] {
    return [...this.landmarks]
  }
}
